ports/gstplugin/servosrc.rs

```diff
@@ -0,0 +1,566 @@
+use gstreamer::subclass::element::ElementClassSubclassExt;
+use gstreamer::subclass::element::ElementImpl;
+use gstreamer::subclass::ElementInstanceStruct;
+use gstreamer::BufferRef;
+use gstreamer::Caps;
+use gstreamer::CoreError;
+use gstreamer::ErrorMessage;
+use gstreamer::FlowError;
+use gstreamer::FlowSuccess;
+use gstreamer::Format;
+use gstreamer::Fraction;
+use gstreamer::FractionRange;
+use gstreamer::IntRange;
+use gstreamer::LoggableError;
+use gstreamer::PadDirection;
+use gstreamer::PadPresence;
+use gstreamer::PadTemplate;
+use gstreamer_base::subclass::base_src::BaseSrcImpl;
+use gstreamer_base::BaseSrc;
+use gstreamer_base::BaseSrcExt;
+use gstreamer_video::VideoFormat;
+use gstreamer_video::VideoFrameRef;
+use gstreamer_video::VideoInfo;
+
+use log::debug;
+use log::info;
+
+use servo::compositing::windowing::AnimationState;
+use servo::compositing::windowing::EmbedderCoordinates;
+use servo::compositing::windowing::EmbedderMethods;
+use servo::compositing::windowing::WindowEvent;
+use servo::compositing::windowing::WindowMethods;
+use servo::embedder_traits::EventLoopWaker;
+use servo::msg::constellation_msg::TopLevelBrowsingContextId;
+use servo::servo_url::ServoUrl;
+use servo::webrender_api::units::DevicePixel;
+use servo::Servo;
+
+use sparkle::gl;
+use sparkle::gl::types::GLuint;
+use sparkle::gl::Gl;
+
+use surfman::platform::generic::universal::context::Context;
+use surfman::platform::generic::universal::device::Device;
+use surfman::SurfaceAccess;
+use surfman::SurfaceType;
+
+use surfman_chains::SwapChain;
+
+use std::cell::RefCell;
+use std::rc::Rc;
+use std::sync::Mutex;
+use std::thread;
+
+pub struct ServoSrc {
+    sender: Sender<ServoSrcMsg>,
+    swap_chain: SwapChain,
+    info: Mutex<Option<VideoInfo>>,
+}
+
+struct ServoSrcGfx {
+    device: Device,
+    context: Context,
+    gl: Rc<Gl>,
+    fbo: GLuint,
+}
+
+impl ServoSrcGfx {
+    fn new() -> ServoSrcGfx {
+        let version = surfman::GLVersion { major: 4, minor: 3 };
+        let flags = surfman::ContextAttributeFlags::empty();
+        let attributes = surfman::ContextAttributes { version, flags };
+
+        let connection = surfman::Connection::new().expect("Failed to create connection");
+        let adapter = surfman::Adapter::default().expect("Failed to create adapter");
+        let mut device =
+            surfman::Device::new(&connection, &adapter).expect("Failed to create device");
+        let descriptor = device
+            .create_context_descriptor(&attributes)
+            .expect("Failed to create descriptor");
+        let context = device
+            .create_context(&descriptor)
+            .expect("Failed to create context");
+        let mut device = Device::Hardware(device);
+        let mut context = Context::Hardware(context);
+        let gl = Gl::gl_fns(gl::ffi_gl::Gl::load_with(|s| {
+            device.get_proc_address(&context, s)
+        }));
+
+        device.make_context_current(&context).unwrap();
```

Comment:
You probably want to make the context not current at the end of this function. It's not guaranteed (and unlikely) that this is the same function as where `fill()` is called later.

Reply:
Done.